Add missing types and interfaces, improve definitions to match 6.6 api…

LKay opened the pull request:
…, remove unused definitions

Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `tsc` without errors.
- [x] Run `npm run lint package-name` if a `tslint.json` is present.

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: http://redux-form.com/6.5.0/docs/api/
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

mhegazy commented:
@omerb  @CarsonF @alex3165 can you please review this change.

janv commented on redux-form/lib/Field.d.ts:
```diff
@@ -202,12 +203,12 @@ interface WrappedFieldInputProps {
      * A function to call when the form field is changed.
      */
-    onChange: FormEventHandler<any>;
+    onChange: (eventOrValue: FormEvent<any> | FieldValue) => void;
```
This should be a `ChangeEvent`, not a `FormEvent`

CarsonF replied:
Also `newValue` and `previousValue` parameters are missing. 

`onFocus`,  `onDragStart`, and `onDrop` props are missing as well.

LKay replied:
This is not really specified in the docs, but I suppose it doesn't hurt to add this one.

CarsonF commented on redux-form/lib/actions.d.ts:
```diff
@@ -138,7 +134,3 @@ export function unregisterField(form: string, name: string): Action;
-
-export function updateSyncErrors(from: string, syncErrors: FormErrors<FormData>, error: any): Action;
-
-export function updateSyncWarnings(form: string, syncWarnings: FormWarnings<FormData>, warning: any): Action;
```
Why are `setSubmitFailed`, `setSubmitSucceeded`, `updateSyncErrors`, and `updateSyncWarnings` removed?

LKay replied:
This is kind of confusing. Documentation says they are no longer present in the public API (http://redux-form.com/6.5.0/docs/api/ActionCreators.md/) but they are still present in the source for 6.5.0 (https://github.com/erikras/redux-form/blob/v6.5.0/src/actions.js). I wanted to be consistent with the official docs in case something gets removed or went to private API (which should be accessed directly).

CarsonF replied:
In my experience with this, it's just missing from docs, not private. I don't see any history for that doc page where the functions are removed, they were just never there.

IMO, it's exported == it's public.

CarsonF commented on redux-form/lib/Field.d.ts:
```diff
@@ -72,15 +66,22 @@ interface BaseFieldProps {
      * does not need a warning, it should return `undefined`.
      */
-    warn?: Validator|Validator[];
+    warn?: Validator;
```
Both `validate` and `warn` can be a list of Validators
http://redux-form.com/6.5.0/examples/fieldLevelValidation/

CarsonF commented on redux-form/lib/FieldArray.d.ts:
```diff
@@ -20,24 +20,23 @@ interface BaseFieldArrayProps {
      * does not need a warning, it should return `undefined`.
      */
-    warn?: Validator|Validator[];
+    warn?: Validator;
```
Both `validate` and `warn` can be a list of Validators
http://redux-form.com/6.5.0/examples/fieldLevelValidation/

CarsonF commented on redux-form/lib/reduxForm.d.ts:
```diff
@@ -33,15 +33,6 @@ export interface Config<FormData extends DataShape, P, S> {
-     *
-     * See Asynchronous Blur Validation Example for more details.
-     */
-    adapter?: (component: string, props: Object) => ReactElement<any>;
```
Missed this, thanks 👍

CarsonF commented on redux-form/lib/reduxForm.d.ts:
```diff
@@ -147,14 +126,13 @@ export interface Config<FormData extends DataShape, P, S> {
+     * component and does not rely on any input or state other than its props and
+     * the selected Redux store's state. Defaults to true.
      */
-    reduxMountPoint?: string;
```
Did I miss this getting removed?

LKay replied:
Looks like this is not part of public API anymore: http://redux-form.com/6.5.0/docs/api/ReduxForm.md/

CarsonF replied:
I can't find it in source either. I know usage was discouraged anyways so I'm fine removing it.

CarsonF commented on redux-form/lib/reduxForm.d.ts:
```diff
@@ -50,14 +41,14 @@ export interface Config<FormData extends DataShape, P, S> {
      * See Asynchronous Blur Validation Example for more details.
      */
-    asyncValidate?(values: FormData, dispatch: Dispatch<S>, props: P, blurredField: string): Promise<any>;
+    asyncValidate?(values: FormData, dispatch: Dispatch<S>, props: P): Promise<any>;
```
`blurredField` is still passed https://github.com/erikras/redux-form/blob/master/src/reduxForm.js#L375

LKay replied:
You're right. Will get this back.

CarsonF commented on redux-form/lib/reduxForm.d.ts:
```diff
@@ -285,65 +268,65 @@ export class SubmissionError<FormData extends DataShape> extends Error {
  * of your decorated form component (outermost component named "ReduxForm").
  */
-export interface Form<FormData extends DataShape, P, S> extends Component<P, any> {
+export interface FormComponent<FormData extends DataShape, P, S> extends ComponentClass<P> {
```
This name change is a BC break, but I see why it's needed.

I'm not sure `FormComponent` is the best name though. The "real" `Form` class is also a component, which makes this ambiguous. Maybe `DecoratedForm` (I'm not sure I like that either)?

Also why make these properties optional?

LKay replied:
Naming this `Form` in the first place was not the best idea. Right now the library exports component named `Form` so the name is in conflict. In this case there is no way to avoid that unfortunately.

CarsonF commented on redux-form/lib/reduxForm.d.ts:
```diff
@@ -285,65 +268,65 @@ export class SubmissionError<FormData extends DataShape> extends Error {
  * Your form component's props can extend this interface.
  */
-export interface FormProps<FormData extends DataShape, P, S> {
+export interface InjectedFormProps<FormData extends DataShape, P, S> {
```
This name change is also a BC break.

Maybe the new `FormProps` could be called something else? How about `FormComponentProps`?

LKay replied:
I will change the name. Anyway this should be probably changed in as soon as v4 is out. I'd rather be consistent with naming so props for component `X` should be named `XProps` such as for `Field`, `Fields` etc.

CarsonF commented on redux-form/lib/Field.d.ts:
```diff
@@ -91,13 +92,13 @@ interface BaseFieldProps {
 export type Parser = (value: FieldValue, name: string) => FieldValue;
 
-export type Validator = (value: FieldValue, allValues?: Object) => undefined | string;
+export type Validator = (value: FieldValue, allValues?: any, props?: any) => any;
```
`allValues` and `props` should really be generics.... I don't think it would be that hard to implement.

LKay replied:
Well, this would be kind of BC, as some other wrapping types have to become generics as well. 

I think we need to go through the definitions sometime as in some places generics are not really needed or could be simplified.

LKay commented on redux-form/lib/Field.d.ts:
```diff
@@ -289,6 +295,17 @@ interface WrappedFieldMetaProps<S> {
+     * A subsequent successful submit will set it back to false.
+     */
+    submitFailed?: boolean;
+
```
Prop `submitSucceeded` seems to be not present in the docs http://redux-form.com/6.5.0/docs/api/Field.md/

CarsonF replied:
Sorry, it's a selector I'm thinking of not a Field prop.

LKay commented:
@CarsonF 
> New getFormNames(state) selector.
Added.

> New getFormInitialValues(state) selector.
Added.

> Allowed multiple form names to be passed to destroy() action creator
Added.

> Added keepSubmitSucceeded config option.
Where did you see that? I don't see that in neither docs or source code.

paulvanbrenk commented:
@CarsonF is this good to go?

CarsonF commented:
>> Added keepSubmitSucceeded config option.

> Where did you see that? I don't see that in neither docs or source code.

https://github.com/erikras/redux-form/pull/2426

CarsonF commented:
These 3 selectors are missing as well
https://github.com/erikras/redux-form/pull/2289

LKay commented:
@CarsonF Allrighty, I will add all these props tomorrow. I have bad experience with some other libraries that don't mention anything in the docs yet export some functions or classes and then remove it without notice so I'm very curious about public (documented) and private API. ;) However, you're right that this library sometimes lack informations in docs.

CarsonF commented:
Sounds good. Feel free to ping Eric and ask him :)

LKay commented:
Ok, I added mentioned selectors and brought back deleted actions. Let me know if there is anything left.

LKay commented:
I noticed `redux-form@6.6.0` just got released so I'll go through the docs and code and will apply additional changes if they are any.

andy-ms commented on redux-form/index.d.ts:
```diff
@@ -1,28 +1,29 @@
+    [P in keyof FormData]?: ReactElement<any> | string;
 } & { _warning?: string };
 
 /**
```
This comment no longer applies, right?

andy-ms commented:
We've recently updated the repository structure.
All packages are now in a `types` subdirectory.
Please rebase this pull request onto the latest `master` branch.

janv commented:
Would be nice if this could be merged now. Anything blocking this?

LKay commented:
@andy-ms Could you quickly review check if there is anything to be changed here to meet comply with recent changes to the file structure and linter updates? If not maybe merge once it already got green light few days ago.

grovesNL commented:
@CarsonF @LKay 

Any thoughts on changing

    /**
     * The current value of the field.
     */
    value: FieldValue;

in Field.d.ts to be a generic type parameter? `FieldValue` is an alias to `any` which isn't useful. Even worse the existing generic type parameter in `WrappedFieldProps` and `WrappedFieldArrayProps` are not used for the same purpose (field uses it for `Dispatch`, field array uses it for `get` etc.).

I don't know if anyone else trying to use `this.props.input.value` now but it doesn't have any type safety and the generic type parameter is misleading when compared to field array.

CarsonF commented:
I would love to make it a generic I'm just not sure how that would happen. Maybe `Pick` or `Record` can be used now?

grovesNL commented:
@CarsonF I meant something the following to use when declaring a component's props (i.e. `React.Component<WrappedFieldProps<{}, MyValueType>>`):

- `interface WrappedFieldProps<S> { input: WrappedFieldInputProps; /* ... */ }`  becomes `interface WrappedFieldProps<S, Value> { input: WrappedFieldInputProps<Value>; }`
- `interface WrappedFieldInputProps { value: FieldValue; }` becomes `interface WrappedFieldInputProps<Value> { value: Value; }`

or instead, `FieldValue` itself could be generic, i.e. `value: FieldValue<Value>`

Ideally the first generic type could be `Value` so it's `WrappedFieldProps<Value, S>` instead, which is more consistent with `WrappedFieldArrayProps`.

grovesNL commented:
@CarsonF As you said, it would be interesting to see if `Pick` or `Record` could be used for prop types in the following example, which is what I think you are talking about:

`<Field name='example' component={MyComponent} a='a value' />`

where the type of prop `a` is taken from `MyComponent`

Although this is a separate issue than I mentioned above.

dumorango commented:
I'm using it this way:

package.json
`"@types/redux-form": "^6.6.3",`

`import { reducer as formReducer } from "redux-form/immutable";`

However, I still get this error:

error TS7016: Could not find a declaration file for module 'redux-form/immutable'. '.../node_modules/redux-form/immutable.js' implicitly has an 'any' type.

I've also tried to change it to:

`import { immutable } from "redux-form";`

and the error turns into:

`node_modules/@types/redux-form/index"' has no exported member 'immutable'.`

Looks like the compiler is correct and there's no "immutable" reference on the 'node_modules/@types/redux-form/index.d.ts' file.

I don't want to turn off the noImplicitAny flag.

I'm new to TypeScript so I'm not sure if I should add the reference to the file and create a for it. 

Thoughts?

Thanks.

andy-ms commented:
@dumorango That's because `redux-form/immutable.d.ts` defined here. You could create a PR adding it.

dumorango commented:
I'll do that @andy-ms. Thanks for the response.
